app/gui2/src/util/ast/abstract.ts

```diff
@@ -126,6 +139,86 @@ export function deleteFromParentBlock(ast: MutableAst) {
     parent.updateLines((lines) => lines.filter((line) => line.expression?.node.id !== ast.id))
 }
 
+/** If the input is a chain of applications of the given left-associative operator, and all the leaves of the
+ *  operator-application tree are identifier expressions, return the identifiers from left to right.
+ *  This is analogous to `ast.code().split(operator)`, but type-enforcing.
+ */
+export function unrollOprChain(
+  ast: Ast,
+  leftAssociativeOperator: string,
+): IdentifierOrOperatorIdentifier[] | null {
+  const idents: IdentifierOrOperatorIdentifier[] = []
+  let ast_: Ast | undefined = ast
+  while (
+    ast_ instanceof OprApp &&
+    ast_.operator.ok &&
+    ast_.operator.value.code() === leftAssociativeOperator
+  ) {
+    if (!(ast_.rhs instanceof Ident)) return null
+    idents.unshift(ast_.rhs.code())
+    ast_ = ast_.lhs
+  }
+  if (!(ast_ instanceof Ident)) return null
+  idents.unshift(ast_.code())
+  return idents
+}
+
+/** If the input is a chain of property accesses (uses of the `.` operator with a syntactic identifier on the RHS), and
+ *  the value at the beginning of the sequence is an identifier expression, return all the identifiers from left to
+ *  right. This is analogous to `ast.code().split('.')`, but type-enforcing.
+ */
+export function unrollPropertyAccess(ast: Ast): IdentifierOrOperatorIdentifier[] | null {
+  const idents: IdentifierOrOperatorIdentifier[] = []
+  let ast_: Ast | undefined = ast
+  while (ast_ instanceof PropertyAccess) {
+    idents.unshift(ast_.rhs.code())
+    ast_ = ast_.lhs
+  }
+  if (!(ast_ instanceof Ident)) return null
+  idents.unshift(ast_.code())
+  return idents
+}
+
+export function parseIdent(ast: Ast): IdentifierOrOperatorIdentifier | null {
+  if (ast instanceof Ident) {
+    return ast.code()
+  } else {
+    return null
+  }
+}
+
+export function parseIdents(ast: Ast): IdentifierOrOperatorIdentifier[] | null {
+  return unrollOprChain(ast, ',')
+}
+
+export function parseQualifiedName(ast: Ast): QualifiedName | null {
+  const idents = unrollPropertyAccess(ast)
+  return idents && normalizeQualifiedName(qnFromSegments(idents))
+}
+
+/* Substitute `name` inside `expression` with `to`. */
+export function substituteQualifiedName(
```

Comment:
This substitution won't work in cases like `Foo.bar (Foo -> Foo.baz)` - here we should replace only one Foo!

I thought a bit, but I don't know how to tackle it. Component Browser Input already analyses its content to know what identifiers are from outside, so it could use the information here also. On the other hand, I'm aware we also want to track import conflicts in other cases than inserting/editing node.

Reply:
As discussed, we would need to refactor CB input logic to work with AST directly for this, and even then it can still be complicated.